tracing: use manual collection in ForkCtxSpan

irfansharif opened the pull request:
The spans created by ForkCtxSpan are expected to outlive the parent, so
automatic collection makes little sense here. It also causes the child
to be linked into the parent, which can delay memory reclamation; some
parent spans are effectively ephemeral.

As an example of this, consider the usage in `replicaDecoder`. The
parent context has a long-lived span, and forks off new tracing spans
for each decoded command. Similar problems occur prop up for the
`raft-worker`/`raftScheduler` goroutines.

---

For tests, we make two changes:
- Logic tests that asserted on async recordings are simply removed,
  we're making a backwards-incompatible change here
- Unit tests that want to assert on trace data from async spans can
  now configure the tracer to include them explicitly

While here, we introduce `tracing.ContextWithRecordingSpanUsing` to
capture the tracer used to construct spans, so most callers can provide
the one available.

Release note (general change): `SHOW TRACE FOR SESSION` previously
included cockroach internal traces for async threads kicked off as part
of user operations. This trace data is no longer captured.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59815)
<!-- Reviewable:end -->

irfansharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59815)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @knz, and @tbg)

___
*[pkg/util/tracing/tracer.go, line 629 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59815#-MSmbheY3Y0f8r_PlMB6:-MSmbheY3Y0f8r_PlMB7:b-1np4aq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b9a022fca813d32512989ba72359ba22398eaf3/pkg/util/tracing/tracer.go#L629)):*
> ```Go
> // the tracer to include recordings from forked/async child spans, when
> // retrieving the recording for a parent span.
> func (t *Tracer) TestingIncludeAsyncSpansInRecordings() {
> ```

I tried introducing an interceptor instead (suggested in https://github.com/cockroachdb/cockroach/pull/59391#pullrequestreview-582170340), and asserting on span recordings on Finish, but it felt too clunky. For tests that want to "capture everything", think this is much easier to work with. It's also only just a handful of tests that want to do this today (the logic tests don't).
___
*[pkg/util/tracing/tracer.go, line 650 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59815#-MSmcL6QAwHvj4xciLw1:-MSmcL6QAwHvj4xciLw2:bqzjojb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b9a022fca813d32512989ba72359ba22398eaf3/pkg/util/tracing/tracer.go#L650)):*
> ```Go
> 		return ctx, nil
> 	}
> 	collectionOpt := WithParentAndManualCollection(sp.Meta())
> ```

Didn't flip the ordering as suggested in https://github.com/cockroachdb/cockroach/pull/59391#pullrequestreview-582170340, the Manual collection will be the one used in everything except for a select few tests, so optimizing for the memory usage seems unnecessary.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Looks like the rabbit hole is going a little deeper still, see the dist sender comments.

> Reviewed 22 of 22 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59815#-:-MT-nbZa1eIkekINuIQc:bhuqgq0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @irfansharif, and @knz)

___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1763 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59815#-MT-fUau35b5YyGGwdX_:-MT-fUau35b5YyGGwdXa:bbgc6bu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1696ce3b878c0fb7e86821367a1c8f1c69995afa/pkg/sql/opt/exec/execbuilder/testdata/select#L1763)):*
> ```
> ----
> querying next range at /Table/73/1/0
> === SPAN START: kv.DistSender: sending partial batch ===
> ```

Unfortunately, DistSender relies on async tasks in more than one place, these are the ones I found (might've missed some, you get the idea):

https://github.com/cockroachdb/cockroach/blob/cdd48a011e849edd404a29b5a8bd52793f019c7d/pkg/kv/kvclient/kvcoord/dist_sender.go#L1370-L1379
https://github.com/cockroachdb/cockroach/blob/cdd48a011e849edd404a29b5a8bd52793f019c7d/pkg/kv/kvclient/kvcoord/dist_sender.go#L993-L1001

The batch responses are combined here:

https://github.com/cockroachdb/cockroach/blob/cdd48a011e849edd404a29b5a8bd52793f019c7d/pkg/kv/kvclient/kvcoord/dist_sender.go#L1221-L1226

At the very least, we need to make sure that `BatchResponse.Combine` appends the spans; otherwise we're losing lots of spans from KV and the whole tracing thing is doomed. It looks like our testing of actual trace histories' accuracy is not very thorough (or rather, does not exist), which we need to fix and might as well now. I think we can set up a test cluster, ten scratch ranges, and a replica eval interceptor that looks at the key for all of the test's puts on each range and emits a metadata containing the key to the trace. The test can then verify that for every put we see the right meta in this range. I *think* it's actually straightforward to sue kvnemesis for this? We execute all regular KV ops in this method:

https://github.com/cockroachdb/cockroach/blob/c955e882ef8181b260a72fe1f08152863ecb1c67/pkg/kv/kvnemesis/applier.go#L135-L140

Each of these should verify that they keys it accesses are reflected in metas that we set up kvnemesis to inject via an eval interceptor. I'm still on the fence whether this is a good first step, randomized testing is great and everything, but it's also good to have a basic manual test of these things that is easier to debug, so maybe that's the better starting point, especially since we expect things to be broken as of this PR.

Not necessarily for now, but something we also need to do:

I think there is also an issue with tracking `Error`s that get retried away internally. These are detached from the batch response here:

https://github.com/cockroachdb/cockroach/blob/cdd48a011e849edd404a29b5a8bd52793f019c7d/pkg/kv/kvclient/kvcoord/dist_sender.go#L1559-L1561

We should make sure they implement `Structured` (do so trivially now) and added to the trace. There's also the question on how to track this error when it occurs:

https://github.com/cockroachdb/cockroach/blob/cdd48a011e849edd404a29b5a8bd52793f019c7d/pkg/kv/kvclient/kvcoord/dist_sender.go#L1547-L1549
___
*[pkg/util/tracing/tracer.go, line 753 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59815#-MT-nMau-KQ3qwsb9TCk:-MT-nMav7hd4rb-bimYr:b-9hkcsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b9a022fca813d32512989ba72359ba22398eaf3/pkg/util/tracing/tracer.go#L753)):*
> ```Go
> // an existing Tracer is unavailable. Tests should typically expect to use
> // ContextWithRecordingSpanUsing instead.
> func ContextWithRecordingSpan(
> ```

Can you nuke this method? All callers in test can just `ContextWithRecordingSpanUsing(ctx, NewTracer(), opName)` and the ~3 callers in prod code should have a tracer they can grab.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59815)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @knz, and @tbg)

___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1763 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59815#-MT-fUau35b5YyGGwdX_:-MT2NlXY9owJ1wPXSolc:b-5jkmwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1696ce3b878c0fb7e86821367a1c8f1c69995afa/pkg/sql/opt/exec/execbuilder/testdata/select#L1763)):*
> At the very least, we need to make sure that BatchResponse.Combine appends the spans; otherwise we're losing lots of spans from KV and the whole tracing thing is doomed

We're already doing that here: https://github.com/cockroachdb/cockroach/blob/9e31dd824b18337a8a7658b0634280cdfd0d7f8a/pkg/roachpb/api.go#L535

But I agree that our testing story here leaves much to be desired. I think that's well deserving of a PR of its own. I need to play around what what that looks like; I'd also prefer being able to isolate and test trace data in unit tests. 

So if I'm reading the above right, given we're already combining spans from batch responses (including those captured using async tasks), does that unblock this PR? All this trace data is later imported into the span at the gateway, which is what we ultimately want (though, yes, would be nice to add tests for it).
___
*[pkg/util/tracing/tracer.go, line 753 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59815#-MT-nMau-KQ3qwsb9TCk:-MT2NkDL85IDoR8uNjxW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b9a022fca813d32512989ba72359ba22398eaf3/pkg/util/tracing/tracer.go#L753)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Can you nuke this method? All callers in test can just `ContextWithRecordingSpanUsing(ctx, NewTracer(), opName)` and the ~3 callers in prod code should have a tracer they can grab.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

angelapwen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> per @tbg :smile_cat: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59815#-:-MTBUI0X5tOrZRkiE10q:b-mj7cw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @tbg)



<!-- Sent from Reviewable.io -->

irfansharif committed:
tracing: use manual collection in ForkCtxSpan

The spans created by ForkCtxSpan are expected to outlive the parent, so
automatic collection makes little sense here. It also causes the child
to be linked into the parent, which can delay memory reclamation; some
parent spans are effectively ephemeral.

As an example of this, consider the usage in `replicaDecoder`. The
parent context has a long-lived span, and forks off new tracing spans
for each decoded command. Similar problems occur prop up for the
`raft-worker`/`raftScheduler` goroutines.

---

For tests, we make two changes:
- Logic tests that asserted on async recordings are simply removed,
  we're making a backwards-incompatible change here
- Unit tests that want to assert on trace data from async spans can
  now configure the tracer to include them explicitly

While here, we introduce `tracing.ContextWithRecordingSpanUsing` to
capture the tracer used to construct spans, so most callers can provide
the one available.

Release note (general change): `SHOW TRACE FOR SESSION` previously
included cockroach internal traces for async threads kicked off as part
of user operations. This trace data is no longer captured.

Co-authored-by: Tobias Grieger <tobias.b.grieger@gmail.com>

irfansharif commented:
Thanks!

bors r+

irfansharif commented:
Skewed with #59760.

bors r+
